tools/java/runfiles/Runfiles.java

```diff
@@ -67,21 +80,84 @@
  */
 public abstract class Runfiles {
 
+  private static final String MAIN_REPOSITORY_NAME = "";
+
+  private static class RepoMappingKey {
+
+    public final String sourceCanonical;
+    public final String targetApparent;
+
+    public RepoMappingKey(String sourceCanonical, String targetApparent) {
+      this.sourceCanonical = sourceCanonical;
+      this.targetApparent = targetApparent;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      }
+      if (o == null || getClass() != o.getClass()) {
+        return false;
+      }
+      RepoMappingKey that = (RepoMappingKey) o;
+      return sourceCanonical.equals(that.sourceCanonical) && targetApparent.equals(
+          that.targetApparent);
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(sourceCanonical, targetApparent);
+    }
+  }
+
+  private final Map<RepoMappingKey, String> repoMapping;
+  private final String sourceCanonical;
+
   // Package-private constructor, so only package-private classes may extend it.
-  private Runfiles() {}
+  private Runfiles(String repoMappingPath, String sourceRepository) throws IOException {
+    this.repoMapping = loadRepositoryMapping(repoMappingPath);
+    this.sourceCanonical = sourceRepository;
+  }
+
+  protected Runfiles(Runfiles runfiles, String sourceRepository) {
+    this.repoMapping = runfiles.repoMapping;
+    this.sourceCanonical = sourceRepository;
+  }
 
   /**
    * Returns a new {@link Runfiles} instance.
    *
+   * <p><strong>Deprecated: With {@code --enable_bzlmod}, this function can only resolve runfiles
+   * correctly if called from the main repository. Use {@link #create(String)} instead. </strong>
```

Comment:
This should always work when used from the main repository, whether with or without Bzlmod. The canonical name of the main repository is the empty string in both cases, so runfiles lookups can succeed from there without any additional information.

Do you see a situation in which this would break? I hope to be able maintain backwards compatibility with the majority of runfiles lookups this way as most of them typically happen from the main repository.

Reply:
Sorry, I didn't understand your last sentence at first. I will clarify that this pattern only works with Bzlmod as long as the repository is part of the root module.